File: archives/C#/0105. Construct Binary Tree from Preorder and Inorder Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode BuildTree(int[] preorder, int[] inorder) {
        if(preorder.Length==0){
            return null;
        }
        
        int i=0;
        while(inorder[i]!=preorder[0]){
            i++;
        }
        
        int lengthK=preorder.Length;
        TreeNode node=new TreeNode(preorder[0]);
        node.left=BuildTree(CutArray(preorder,1,i),CutArray(inorder,0,i));
        node.right=BuildTree(CutArray(preorder,i+1,lengthK-i-1),CutArray(inorder,i+1,lengthK-i-1));
        return node;
    }
    
    public int[] CutArray(int[] rootorder,int start,int cutLength){
        List<int> rep=new List<int>();
        for(int i=0;i<cutLength;i++){
            rep.Add(rootorder[start+i]);
        }
        return rep.ToArray();
    }
}

File: archives/leetcode2/0105. Construct Binary Tree from Preorder and Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: 'List[int]', inorder: 'List[int]') -> 'TreeNode':
        if not preorder:
            return None
        k=0
        while inorder[k]!=preorder[0]:
            k+=1
        root=TreeNode(preorder[0])
        root.left=self.buildTree(preorder[1:k+1],inorder[:k])
        root.right=self.buildTree(preorder[k+1:],inorder[k+1:])
        return root
